PEP 810: Explicit lazy imports (#4622)
python/peps: peps/pep-0810.rst

        Brittany Reynoso <brittanyrey@gmail.com>,
        Noah Kim <noahbkim@gmail.com>,
        Tim Stumbaugh <me@tjstum.com>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Created: 02-Oct-2025
Python-Version: 3.15

Abstract
========

This PEP introduces lazy imports as an explicit language feature. Lazy imports
defer the loading and execution of a module until the first time the imported
name is used, rather than at the point of the import statement.

By allowing developers to mark individual imports as lazy with explicit syntax, Python programs
can reduce startup time, memory usage, and unnecessary work. This is
particularly beneficial for command-line tools, test suites, and applications
with large dependency graphs.

The proposal preserves full backwards compatibility: normal import statements
remain unchanged, and lazy imports are enabled only where explicitly requested.

Motivation
==========

A common convention in Python code is to place all imports at the module
level, typically at the beginning of the file. This avoids repetition, makes dependencies clear
and minimizes runtime overhead by only evaluating an import statement once

> per module. A major drawback with this approach is that importing the first
> module for an execution of Python (the "main" module) often triggers an immediate
> cascade of imports, loading many dependencies that may never be used. The effect
> is especially costly for command-line tools with multiple subcommands, where
> even running the command with ``--help`` can load dozens of unnecessary modules and
> take several seconds, just to get feedback to the user of how to run the program at
> all. Then the user incurs all that overhead again when they figure out the command
> they want and invoke the program "for real."

Reviewer: ```suggestion
per module.

A major drawback with this approach is that importing the first
module for an execution of Python (the "main" module) often triggers an immediate
cascade of imports, and optimistically loads many dependencies that may never be used. The effect
is especially costly for command-line tools with multiple subcommands, where
even running the command with ``--help`` can load dozens of unnecessary modules and
take several seconds. This basic example demonstrates what must be loaded just to get helpful feedback to the user on how to run the program at
all. Inefficiently, the user incurs this overhead again when they figure out the command
they want and invoke the program "for real."
```

> A somewhat common response is to delay imports by moving them into functions
> (inline imports), but this practice is very manual to implement and maintain.
> Additionally, it obfuscates the full set of dependencies for a module. Analysis

Reviewer: ```suggestion
A somewhat common way to delay imports is to move the imports into functions
(inline imports), but this practice is very manual to implement and maintain.
Additionally, it obfuscates the full set of dependencies for a module. Analysis
```

of the Python standard library shows that approximately 17% of all imports outside
tests (nearly 3500 total imports across 730 files) are already placed inside
functions, classes, or methods specifically to defer their execution. This
demonstrates that developers are already manually implementing lazy imports in
performance-sensitive code, but doing so requires scattering imports throughout
the codebase and makes the full dependency graph harder to understand at a
glance.

> The standard library provides ``importlib.util.LazyLoader`` to solve some of these

Reviewer: ```suggestion
The standard library provides ``importlib.util.LazyLoader`` to solve some of these inefficiency
```

problems. It permits imports at the module level to work *mostly* like inline
imports do. Scientific Python libraries have adopted a similar pattern, formalized
in `SPEC 1 <https://scientific-python.org/specs/spec-0001/>`_. There's also the
third-party `lazy_loader <https://pypi.org/project/lazy-loader/>`_ package.
Imports used solely for static type checking are another source of potentially unneeded
imports, and there are similarly disparate approaches to minimizing the overhead.

> These approaches don't cover all cases, add runtime overhead in unexpected places,
> are non-obvious, and non-standardized.

Reviewer: ```suggestion
All use cases are not covered by these approaches; however, these approaches add runtime overhead in unexpected places,
in non-obvious ways, and without a clear standard.
```

PEP: 810
Title: Explicit lazy imports
Author: Pablo Galindo <pablogsal@python.org>,
        Germán Méndez Bravo <german.mb@gmail.com>,
        Thomas Wouters <thomas@python.org>,
        Dino Viehland <dinoviehland@gmail.com>,
        Brittany Reynoso <brittanyrey@gmail.com>,
        Noah Kim <noahbkim@gmail.com>,
        Tim Stumbaugh <me@tjstum.com>
Discussions-To: Pending
Status: Draft
Type: Standards Track
Topic: Release

> Created: 22-09-2025

Reviewer 2: This is the date of PEP number assignment.
```suggestion
Created: 02-Oct-2025
```

are always eager.

Examples of syntax errors:

.. code-block:: python

  # SyntaxError: lazy import not allowed inside functions
  def foo():
      lazy import json

  # SyntaxError: lazy import not allowed inside classes
  class Bar:
      lazy import json

  # SyntaxError: lazy import not allowed inside try/except blocks
  try:
      lazy import json
  except ImportError:
      pass

  # SyntaxError: lazy import not allowed inside with blocks
  with suppress(ImportError):
      lazy import json

  # SyntaxError: lazy from ... import * is not allowed
  lazy from json import *

Semantics
---------

When the ``lazy`` keyword is used, the import becomes *potentially lazy*.
Unless lazy imports are disabled or suppressed (see below), the module is
not loaded immediately at the import statement; instead, a lazy proxy object
is created and bound to the name. The actual module is loaded on first use
of that name.

Example:

.. code-block:: python

  import sys

  lazy import json

  print('json' in sys.modules)  # False - module not loaded yet

  # First use triggers loading
  result = json.dumps({"hello": "world"})

  print('json' in sys.modules)  # True - now loaded

> A module may contain a ``__lazy_modules__`` attribute, which is a list of

Reviewer 3: How does this interact with from imports? If I have `__lazy_modules__ = ["a"]`, does that make both `from a import b` and `from a import c` automatically lazy? How about `import a.b` or `from a.b import c`?

Also, does it have to be a list or can it be any sequence?

object, and if so, returns a lazy object for each name instead.

The end result of this process is that lazy imports (regardless of how they
are enabled) result in lazy objects being assigned to global variables.

Lazy module objects do not appear in ``sys.modules``, they're just listed in
the ``sys.lazy_modules`` set. Under normal operation lazy objects should
only end up stored in global variables, and the common ways to access those
variables (regular variable access, module attributes) will resolve lazy
imports ("reify") and replace them when they're accessed.

It is still possible to expose lazy objects through other means, like
debuggers. This is not considered a problem.

Reification
-----------

When a lazy object is first used, it needs to be reified. This means
resolving the import at that point in the program and replacing the lazy
object with the concrete one. Reification imports the module in the same way
as it would have been if it had been imported eagerly, barring intervening
changes to the import system (e.g. to ``sys.path``, ``sys.meta_path``,
``sys.path_hooks`` or ``__import__``).

Reification still calls ``__import__`` to resolve the import. Once the

> module is reified, it's removed from ``sys.lazy_modules``. When a package is

Reviewer 3: Even if it's also lazily imported somewhere else? Text below suggests that reification doesn't affect other instances of the lazy module that may live elsewhere.

Reviewer 4: Yes, it's removed when the import becomes concrete (when it's loaded and added to `sys.modules`), not when all lazy objects are reified (which may never happen). It's the set of module names that are currently *not* loaded but have been lazily imported. I've clarified the text.

still returns lazy proxy objects that need to be manually reified upon use.
A lazy object can be resolved explicitly by calling the ``get`` method.
Other, more indirect ways of accessing arbitrary globals (e.g. inspecting
``frame.f_globals``) also do **not** reify all the objects.

Example using ``globals()``:

.. code-block:: python

  import sys
  lazy import json

  # Calling globals() does NOT trigger reification
  g = globals()

  print('json' in sys.modules)  # False - still lazy
  print(type(g['json']))  # <class 'lazy_import'>

  # Explicitly reify using the get() method
  resolved = g['json'].get()

  print(type(resolved))  # <class 'module'>
  print('json' in sys.modules)  # True - now loaded

Bytecode and adaptive specialization
-------------------------------------

Lazy imports are implemented through modifications to four bytecode instructions:
``IMPORT_NAME``, ``IMPORT_FROM``, ``LOAD_GLOBAL``, and ``LOAD_NAME``.

The ``lazy`` syntax sets a flag in the ``IMPORT_NAME`` instruction's oparg
(``oparg & 0x01``). The interpreter checks this flag and calls
``_PyEval_LazyImportName()`` instead of ``_PyEval_ImportName()``, creating a lazy

> import object rather than executing the import immediately. The ``IMPORT_FROM``

Reviewer 3: Why do import and import from use a different mechanism?

And how does this interact with the `__lazy_modules__` global? Is that syntactic (i.e. it affects the bytecode) or does it take effect at runtime?

Also, the discussion of bytecode and specific C API functions doesn't belong here: the Specification should specify the user-facing behavior, not implementation details in CPython.

Thread-safety and reification
-----------------------------

First use of a lazy binding follows the existing import-lock discipline. Exactly
one thread performs the import and **atomically rebinds** the importing module's
global to the resolved object. Concurrent readers thereafter observe the real
object.

Lazy imports are thread-safe and have no special considerations for free-threading.
A module that would normally be imported in the main thread may be imported in a
different thread if that thread triggers the first access to the lazy import. This
is not a problem: the import lock ensures thread safety regardless of which thread
performs the import.

Subinterpreters are supported. Each subinterpreter maintains its own
``sys.lazy_modules`` and import state, so lazy imports in one subinterpreter do
not affect others.

Typing and tools
----------------

Type checkers and static analyzers may treat ``lazy`` imports as ordinary
imports for name resolution. At runtime, annotation-only imports can be marked
``lazy`` to avoid startup overhead. IDEs and debuggers should be prepared to
display lazy proxies before first use and the real objects thereafter.

> Lazy imports filter

Reviewer 3: This is in the "Backwards Compatibility" section but shouldn't be, since it proposes a new feature. Move it to Specification.

-------------------

This PEP adds two new functions to the ``sys`` module to manage the lazy imports filter:

* ``sys.set_lazy_imports_filter(func)`` - Sets the filter function. The ``func``

>   parameter must have the signature: ``func(importer: str, name: str, fromlist: tuple[str, ...] | None) -> bool``

Reviewer 3: What does this do? The example suggests that if the function returns false for an import, it's not lazily imported, but this should be said explicitly.

Reviewer 4: It's mentioned in the spec, but I've clarified it in this part too.

* Always import needed submodules explicitly. It is not enough to rely on a different import
  to ensure a module has its submodules as attributes. Plainly, unless there is an
  explicit ``from . import bar`` in ``foo/__init__.py``, always use ``import
  foo.bar; foo.bar.Baz``, not ``import foo; foo.bar.Baz``. The latter only works
  (unreliably) because the attribute ``foo.bar`` is added as a side effect of
  ``foo.bar`` being imported somewhere else.

* Users who are moving imports into functions to improve startup time, should instead
  consider keeping them where they are but adding the ``lazy`` keyword. This allows
  them to keep dependencies clear and avoid the overhead of repeatedly re-resolving
  the import but will still speed up the program.

  .. code-block:: python

    # Before: Inline import (repeated overhead)
    def process_data(data):
        import json  # Re-resolved on every call
        return json.dumps(data)

    # After: Lazy import at module level
    lazy import json

    def process_data(data):
        return json.dumps(data)  # Loaded once on first call

* Avoid using wild card (star) imports, as those are always eager.

FAQ
===

> **Q: How does this differ from the rejected PEP 690?**

Reviewer 3: ```suggestion
**Q: How does this differ from the rejected :pep:`690`?**
```

execute only part of a ``.py`` file. When you first access ``Class``, Python:

1. Loads and executes the entire ``module.py`` file
2. Extracts the ``Class`` attribute from the resulting module object
3. Binds ``Class`` to the name in your namespace

This is identical to eager ``from module import Class`` behavior. The only difference
with lazy imports is that steps 1-3 happen on first use instead of at the import
statement.

.. code-block:: python

  # heavy_module.py
  print("Loading heavy_module")  # This ALWAYS runs when module loads

  class MyClass:
      pass

  class UnusedClass:
      pass  # Also gets defined, even though we don't import it

  # app.py
  lazy from heavy_module import MyClass

  print("Import statement done")  # heavy_module not loaded yet
  obj = MyClass()                  # NOW "Loading heavy_module" prints
                                   # (and UnusedClass gets defined too)

**Key point**: Lazy imports defer *when* a module loads, not *what* gets loaded.
You cannot selectively load only parts of a module—Python's import system doesn't
support partial module execution.

**Q: What about type annotations and** ``TYPE_CHECKING`` **imports?**

> A: Lazy imports eliminate the need for ``TYPE_CHECKING`` guards. You can write:

Reviewer 3: Not in all cases; not all `TYPE_CHECKING` uses are related to imports.

A: Wild card (star) imports cannot be lazy - they remain eager. This is because the
set of names being imported cannot be determined without loading the module. Using the
``lazy`` keyword with star imports will be a syntax error. If lazy imports are globally
enabled, star imports will still be eager.

**Q: How do lazy imports interact with import hooks and custom loaders?**

A: Import hooks and loaders work normally. When a lazy object is first used, the
standard import protocol runs, including any custom hooks or loaders that were
in place at reification time.

**Q: What happens in multi-threaded environments?**

A: Lazy import reification is thread-safe. Only one thread will perform the
actual import, and the binding is atomically updated. Other threads will see
either the lazy proxy or the final resolved object.

**Q: Can I force reification of a lazy import without using it?**

A: Yes, accessing a module's ``__dict__`` will reify all lazy objects in that
module. Individual lazy objects can be resolved by calling their ``get()`` method.

**Q: What's the difference between** ``globals()`` **and** ``mod.__dict__`` **for lazy imports?**

> A: Calling ``globals()`` returns the module's dictionary without reifying lazy

Reviewer 3: I guess this can be discussed on Discourse but I'm not sure I see why these two should be different.

.. code-block:: python

  # In your module:
  lazy import json

  g = globals()
  print(type(g['json']))  # <class 'lazy_import'> - your problem

  # From external code:
  import sys
  mod = sys.modules['your_module']
  d = mod.__dict__
  print(type(d['json']))  # <class 'module'> - reified for external access

This distinction means adding lazy imports and calling ``globals()`` is your
responsibility to manage, while external code accessing ``mod.__dict__`` always
sees fully loaded modules.

**Q: Why not use** ``importlib.util.LazyLoader`` **instead?**

A: ``LazyLoader`` has significant limitations:

- Requires verbose setup code for each lazy import
- Has ongoing performance overhead on every attribute access
- Doesn't work well with ``from ... import`` statements
- Less clear and standard than dedicated syntax

**Q: Will this break tools like** ``isort`` **or** ``black``?

A: Tools will need updates to recognize the ``lazy`` keyword, but the changes
should be minimal since the import structure remains the same. The keyword
appears at the beginning, making it easy to parse.

**Q: How do I know if a library is compatible with lazy imports?**

> A: Most libraries should work fine with lazy imports. Libraries that might have issues:

Reviewer 3: This seems most of an issue with the flag that enables lazy imports globally, for what it's worth. If you just put lazy imports in your code using the library, it's very difficult for it to break a third-party library.

Reviewer 4: That's unfortunately not true, it's a potential problem even with explicit lazy imports.

- Those with essential import-time side effects (registration, monkey-patching)
- Those that expect specific import ordering
- Those that modify global state during import

When in doubt, test lazy imports with your specific use cases.

**Q: Can I use lazy imports inside functions?**

A: No, the ``lazy`` keyword is only allowed at module level. For function-level
lazy loading, use traditional inline imports or move the import to module level
with ``lazy``.

**Q: What about forwards compatibility with older Python versions?**

A: Use the ``__lazy_modules__`` global for compatibility:

.. code-block:: python

  # Works on Python 3.15+ as lazy, eager on older versions
  __lazy_modules__ = ['expensive_module', 'expensive_module_2']
  import expensive_module
  from expensive_module_2 import MyClass

> The ``__lazy_modules__`` attribute is a list of module name strings. When an import

Reviewer 3: This is much more detail than before; it belongs in the specification.

statement is executed, Python checks if the module name being imported appears in
``__lazy_modules__``. If it does, the import is treated as if it had the ``lazy``
keyword (becoming *potentially lazy*). On Python versions before 3.15 that don't
support lazy imports, the ``__lazy_modules__`` attribute is simply ignored and
imports proceed eagerly as normal.

This provides a migration path until you can rely on the ``lazy`` keyword. For
maximum predictability, it's recommended to define ``__lazy_modules__`` once,
before any imports. But as it is checked on each import, it can be modified between
``import`` statements.

**Q: How do explicit lazy imports interact with PEP-649/PEP-749**

> A: If an annotation is not stringified, then it is a reference to the type, but it

Reviewer 3: Not sure what "the type" is here. I'd suggest something like:
```suggestion
A: If an annotation contains a reference to a name that was imported lazily, it
```

will only be resolved if the annotation is accessed. In the example below, the ``fake_typing`` module
is only loaded when the user inspects the ``__annotations__`` dictionary. The ``fake_typing``

> module would also be loaded if the user had leveraged ``inspect.get_annotations()`` or ``getattr`` to access

Reviewer 3: ```suggestion
module would also be loaded if the user had leveraged ``annotationlib.get_annotations()`` or ``getattr`` to access
```

Reviewer 4: Suggestion incorporated.

the annotations.

.. code-block:: python

  lazy from fake_typing import MyFakeType
  def foo(x: MyFakeType):
    pass
  print(foo.__annotations__)  # Triggers loading the fake_typing module

**Q: How do lazy imports interact with** ``dir()``, ``getattr()``, **and module introspection?**

A: Accessing lazy imports through normal attribute access or ``getattr()`` will trigger

> reification. Calling ``dir()`` on a module will reify all lazy imports in that module

Reviewer 3: Why? dir() only shows the names in the namespace, not their values, so reification doesn't change anything.

to ensure the directory listing is complete. This is similar to accessing ``mod.__dict__``.

.. code-block:: python

  lazy import json

  # Before any access
  # json not in sys.modules

  # This triggers reification:

>   dumps_func = getattr(json, 'dumps')

Reviewer 3: Seems to me this example has nothing to do with getattr(), the reification happens because you access the `json` global. Should this example instead show a getattr in a different module?

Author: Maybe we can do `print(json)` as an example of how to trigger reification?

object to support lazy imports. We recognize that this data structure is highly tuned, heavily used
across the codebase, and very performance sensitive. Because of the importance of this data structure
and the desire to keep the implementation of lazy imports encapsulated from users who may have no
interest in the feature, we've decided to invest in an alternate approach.

The dictionary is the foundational data structure in Python. Every object's attributes are stored
in a dict, and dicts are used throughout the runtime for namespaces, keyword arguments, and more.
Adding any kind of hook or special behavior to dicts to support lazy imports would:

1. Prevent critical interpreter optimizations including future JIT compilation
2. Add complexity to a data structure that must remain simple and fast
3. Affect every part of Python, not just import behavior
4. Violate separation of concerns—the hash table shouldn't know about the import system

Past decisions that violated this principle of keeping core abstractions clean have caused
significant pain in the CPython ecosystem, making optimization difficult and introducing
subtle bugs.

Support of Import Cycles
------------------------

> The initial PEP for lazy imports (PEP 690) contained support for eliminating most import cycles,

Reviewer 4: I think this is a remnant from when we tried to make import cycles with lazy components behave more like they do without lazy imports, but it turned out to be more difficult than we thought, so there's not that much difference with PEP 690 now. Removed it.

but this meant that in some cases a codebase would no longer run successfully after disabling
lazy imports. For the explicit lazy imports PEP we’ve decided not to add support for this use case.
While codebases with heavily nested dependencies may benefit from this feature, supporting import
cycles runs the risk of creating a dependency on lazy imports. As the intention of this PEP is to
give users the power to opt in or opt out of lazy imports based on the specific needs within their
codebases, we’ve decided not to implement this feature and instead prioritize backwards compatibility.

Placing the ``lazy`` Keyword in the Middle of From Imports
----------------------------------------------------------

While we found ``from foo lazy import bar`` to be a really intuitive placement for the new explicit syntax,
we quickly learned that placing the ``lazy`` keyword here is already syntactically allowed in Python. This

> is because ``from foo. lazy import bar`` is legal syntax.

Author: I think this example should be `from .lazy import` as that's the one is really a problem

> is because ``from foo. lazy import bar`` is legal syntax.

Reviewer 4: It all comes down to "whitespace does not matter", so I've added that clarification.

PEP: 810
Title: Explicit lazy imports
Author: Pablo Galindo <pablogsal@python.org>,
        Germán Méndez Bravo <german.mb@gmail.com>,
        Thomas Wouters <thomas@python.org>,
        Dino Viehland <dinoviehland@gmail.com>,
        Brittany Reynoso <brittanyrey@gmail.com>,
        Noah Kim <noahbkim@gmail.com>,
        Tim Stumbaugh <me@tjstum.com>

> Discussions-To: https://discuss.python.org/t/pep-810-explicit-lazy-imports/104131

Reviewer 4: The template should be adjusted then :) It says "Right before or after initial merging" and we were assuming we'd merge soon, since the PEP will undoubtedly go through multiple revisions before being sent to the SC anyway.

Status: Draft
Type: Standards Track
Created: 02-Oct-2025
Python-Version: 3.15

Abstract

> ========

Reviewer 4: Agreed, done.

and minimizes runtime overhead by only evaluating an import statement once
per module.

A major drawback with this approach is that importing the first
module for an execution of Python (the "main" module) often triggers an immediate
cascade of imports, and optimistically loads many dependencies that may never be used. The effect
is especially costly for command-line tools with multiple subcommands, where
even running the command with ``--help`` can load dozens of unnecessary modules and
take several seconds. This basic example demonstrates what must be loaded just to get helpful feedback to the user on how to run the program at
all. Inefficiently, the user incurs this overhead again when they figure out the command
they want and invoke the program "for real."

A somewhat common way to delay imports is to move the imports into functions
(inline imports), but this practice is very manual to implement and maintain.
Additionally, it obfuscates the full set of dependencies for a module. Analysis
of the Python standard library shows that approximately 17% of all imports outside
tests (nearly 3500 total imports across 730 files) are already placed inside

> functions, classes, or methods specifically to defer their execution. This

Reviewer 4: Indeed (although we don't want to support lazy import in class bodies, it's not because they're already deferred, but because the variable semantics are different.)

the codebase and makes the full dependency graph harder to understand at a
glance.

The standard library provides ``importlib.util.LazyLoader`` to solve some of these inefficiency
problems. It permits imports at the module level to work *mostly* like inline
imports do. Scientific Python libraries have adopted a similar pattern, formalized
in `SPEC 1 <https://scientific-python.org/specs/spec-0001/>`_. There's also the
third-party `lazy_loader <https://pypi.org/project/lazy-loader/>`_ package.
Imports used solely for static type checking are another source of potentially unneeded
imports, and there are similarly disparate approaches to minimizing the overhead.
All use cases are not covered by these approaches; however, these approaches add runtime overhead in unexpected places,
in non-obvious ways, and without a clear standard.

This proposal introduces **lazy imports syntax** with a design that is local, explicit,
controlled, and granular. Each of these qualities is essential to making the feature
predictable and safe to use in practice.

The behavior is **local**: laziness applies only to the specific import marked

> with the ``lazy`` keyword, and it does not cascade recursively into other

Reviewer 4: Well, these four aspects are very motivational to the design. They are not just aspects of the design, they're also how the idea differentiates itself from PEP 690.

Lazy imports are **controlled**, in the sense that deferred loading is only
triggered by the importing code itself. In the general case, a library will only
experience lazy imports if its own authors choose to mark them as such. This
avoids shifting responsibility onto downstream users and prevents accidental
surprises in library behavior. Since library authors typically manage their own
import subgraphs, they retain predictable control over when and how laziness is
applied.

The mechanism is also **granular**. It is introduced through explicit syntax on
individual imports, rather than a global flag or implicit setting. This allows
developers to adopt it incrementally, starting with the most
performance-sensitive areas of a codebase. As this feature is introduced to the
community, we want to make the experience of onboarding optional, progressive, and
adaptable to the needs of each project.

In addition to the new lazy import syntax, we *also* propose a way to
control lazy imports at the application level: globally disabling or
enabling, and selectively disabling. These are provided for debugging,
testing and experimentation, and are not expected to be the common way to

> control lazy imports.

Reviewer 4: The entire sentence is talking about controlling lazy imports, so I would think the antecedent is obvious: It's about globally disabling lazy imports, globally enabling lazy imports, and selectively disabling lazy imports. I'll see if I can reword it further to make it more obvious.

> The design of lazy imports provides several concrete advantages:

Reviewer 4: I've removed the part about the global flag from the motivation (it already existed in the rationale), and reworded the concrete advantages so they talk about the *concept* of lazy imports, rather than *this specific implementation*. I believe that fits better in the Motivation / Rationale distinction.

* Command-line tools are often invoked directly by a user, so latency — in particular

>   startup latency — is quite noticeable. These programs are also typically

Reviewer 4: Fixed throughout the doc.

  short-lived processes (contrasted with, e.g., a web server). Most conventions
  would have a CLI with multiple subcommands import every dependency up front,

>   even if the user only requests ``tool --help`` (or ``tool subcommand --help``).

Reviewer 4: Agreed, removed.

semantics of the proposal. If the community expresses a strong preference for a
different technical path that still preserves the same core semantics or there
is fundamental disagreement over the specific option, we have included the
brainstorming we have already completed in preparation for this proposal as reference.

The choice to introduce a new ``lazy`` keyword reflects the need for explicit
syntax. Import behavior is too fundamental to be left implicit or hidden behind
global flags or environment variables. By marking laziness directly at the
import site, the intent is immediately visible to both readers and tools. This
avoids surprises, reduces the cognitive burden of reasoning about imports, and
keeps the semantics in line with Python's tradition of explicitness.

Another important decision is to represent lazy imports with proxy objects in
the module's namespace, rather than by modifying dictionary lookup. Earlier
approaches experimented with embedding laziness into dictionaries, but this
blurred abstractions and risked affecting unrelated parts of the runtime. The
dictionary is a fundamental data structure in Python—literally every object is

> built on top of dicts—and adding hooks to dictionaries would prevent critical

Reviewer 4: Replaced with `--` instead as part of the global search-and-replace.

optimizations and complicate the entire runtime. The proxy approach is simpler:
it behaves like a placeholder until first use, at which point it resolves the
import and rebinds the name. From then on, the binding is indistinguishable
from a normal import. This makes the mechanism easy to explain and keeps the
rest of the interpreter unchanged.

Compatibility for library authors was also a key concern. Many maintainers need
a migration path that allows them to support both new and old versions of

> Python at once. For this reason, the proposal includes the ``__lazy_modules__``

Reviewer 4: Presumably this should happen for all occurrences of ``__lazy_modules__``? Done.

global as a transitional mechanism. A module can declare which imports should
be treated as lazy (by listing the module names as strings), and on Python 3.15
or later those imports will become lazy automatically. On earlier versions the
declaration is ignored, leaving imports eager. This gives authors a practical
bridge until they can rely on the keyword as the canonical syntax.

Finally, the feature is designed to be adopted incrementally. Nothing changes
unless a developer explicitly opts in, and adoption can begin with just a few
imports in performance-sensitive areas. This mirrors the experience of gradual
typing in Python: a mechanism that can be introduced progressively, without
forcing projects to commit globally from day one. Notably, the adoption can also
be done from the "outside in," permitting CLI authors to introduce lazy imports
and speed up user-facing tools, without requiring changes to every library the
tool might use.

By combining explicit syntax, a simple runtime model, a compatibility layer,
and gradual adoption, this proposal balances performance improvements with the

> clarity and stability that Python users expect.

Reviewer 4: Done.

are always eager.

Examples of syntax errors:

.. code-block:: python

  # SyntaxError: lazy import not allowed inside functions
  def foo():
      lazy import json

  # SyntaxError: lazy import not allowed inside classes
  class Bar:
      lazy import json

  # SyntaxError: lazy import not allowed inside try/except blocks
  try:
      lazy import json
  except ImportError:
      pass

  # SyntaxError: lazy import not allowed inside with blocks
  with suppress(ImportError):
      lazy import json

  # SyntaxError: lazy from ... import * is not allowed
  lazy from json import *

Semantics
---------

When the ``lazy`` keyword is used, the import becomes *potentially lazy*.
Unless lazy imports are disabled or suppressed (see below), the module is
not loaded immediately at the import statement; instead, a lazy proxy object
is created and bound to the name. The actual module is loaded on first use
of that name.

Example:

.. code-block:: python

  import sys

  lazy import json

  print('json' in sys.modules)  # False - module not loaded yet

  # First use triggers loading
  result = json.dumps({"hello": "world"})

  print('json' in sys.modules)  # True - now loaded

> A module may contain a ``__lazy_modules__`` attribute, which is a sequence of

Reviewer 4: It can be anywhere -- it's not a syntactic construct -- but as is mentioned elsewhere in the PEP, it's looked up at runtime, so it won't affect imports that happen before it's assigned. I think that should be apparent from the spec.

  print('json' in sys.modules)  # False - still lazy

  # main.py
  import sys
  import my_module

  # Accessing __dict__ from external code DOES reify all lazy imports
  d = my_module.__dict__

  print('json' in sys.modules)  # True - reified by __dict__ access
  print(type(d['json']))  # <class 'module'>

However, calling ``globals()`` does **not** trigger reification — it returns
the module's dictionary, and accessing lazy objects through that dictionary
still returns lazy proxy objects that need to be manually reified upon use.
A lazy object can be resolved explicitly by calling the ``get`` method.
Other, more indirect ways of accessing arbitrary globals (e.g. inspecting
``frame.f_globals``) also do **not** reify all the objects.

Example using ``globals()``:

.. code-block:: python

  import sys
  lazy import json

  # Calling globals() does NOT trigger reification
  g = globals()

  print('json' in sys.modules)  # False - still lazy
  print(type(g['json']))  # <class 'lazy_import'>

  # Explicitly reify using the get() method
  resolved = g['json'].get()

  print(type(resolved))  # <class 'module'>
  print('json' in sys.modules)  # True - now loaded

Implementation

> ==============

Reviewer 4: Done.

A reference implementation is available at:
https://github.com/LazyImportsCabal/cpython/tree/lazy

Alternate Implementation Ideas
==============================

Here are some alternative design decisions that were considered during the development
of this PEP. While the current proposal represents what we believe to be the best balance
of simplicity, performance, and maintainability, these alternatives offer different
trade-offs that may be valuable for implementers to consider or for future refinements.

Leveraging a Subclass of Dict
-----------------------------

Instead of updating the internal dict object to directly add the fields needed to support lazy imports,
we could create a subclass of the dict object to be used specifically for Lazy Import enablement. This
would still be a leaky abstraction though - methods can be called directly such as ``dict.__getitem__``
and it would impact the performance of globals lookup in the interpreter.

Alternate Keyword Names
-----------------------

For this PEP, we decided to propose ``lazy`` for the explicit keyword as it felt the most familar to those
already focused on optimizing import overhead. We also considered a variety of other

> options to support explicit lazy imports. The most compelling alternates were ``defer`` and ``delay``.

Reviewer 4: Not sure if this was a genuine comment or not, but if so, that's a topic for the discuss thread :)

in a dict, and dicts are used throughout the runtime for namespaces, keyword arguments, and more.
Adding any kind of hook or special behavior to dicts to support lazy imports would:

1. Prevent critical interpreter optimizations including future JIT compilation
2. Add complexity to a data structure that must remain simple and fast
3. Affect every part of Python, not just import behavior
4. Violate separation of concerns—the hash table shouldn't know about the import system

Past decisions that violated this principle of keeping core abstractions clean have caused
significant pain in the CPython ecosystem, making optimization difficult and introducing
subtle bugs.

Placing the ``lazy`` Keyword in the Middle of From Imports
----------------------------------------------------------

While we found ``from foo lazy import bar`` to be a really intuitive placement for the new explicit syntax,
we quickly learned that placing the ``lazy`` keyword here is already syntactically allowed in Python. This
is because ``from . lazy import bar`` is legal syntax (because whitespace
does not matter.)

Placing the ``lazy`` Keyword at the End of Import Statements

> ------------------------------------------------------------

Reviewer 4: Fixed.
